pkg/config/tlscfg/options.go

```diff
@@ -114,7 +114,6 @@ func (p *Options) Config(logger *zap.Logger) (*tls.Config, error) {
 		}
 	}
 
-	go p.certWatcher.watchChangesLoop(tlsCfg.RootCAs, tlsCfg.ClientCAs)
 	return tlsCfg, nil
 }
 
```

Comment:
GH doesn't allow to comment on L142, but it's causing a race condition in the CI.
I think we could use the same mutex in certWatcher to serialize access to the cert pools.
```
WARNING: DATA RACE
Read at 0x00c0000105e8 by goroutine 22:
  crypto/x509.(*CertPool).addCertFunc()
      /opt/hostedtoolcache/go/1.20.2/x64/src/crypto/x509/cert_pool.go:189 +0x6b8
  crypto/x509.(*CertPool).AppendCertsFromPEM()
      /opt/hostedtoolcache/go/1.20.2/x64/src/crypto/x509/cert_pool.go:227 +0x4f1
  github.com/jaegertracing/jaeger/pkg/config/tlscfg.addCertToPool()
      /home/runner/work/jaeger/jaeger/pkg/config/tlscfg/options.go:142 +0x1d7
 ```

Reply:
I added a mutex, but I think there is a bigger issue with race conditions, which is outside the scope of this PR. Namely, the race condition showed that certPools themselves are not safe for concurrent access. The mutex I added prevents competing writes, but it's not clear to me what that means to the actual RPC layers that utilize this TLS configuration. It's possible that they only read the certPool at the beginning, or perhaps at rare re-connection events. But if they do read pools from time to time, this is a clear race condition with use trying to update those pools. I don't really see how this can be solved in Go standard tls package.

Reply:
@tsaarni I wonder if you ran into this race condition

Reply:
@yurishkuro I have not bumped into it, but we have refrained from rotating CA certificates at runtime. 

I agree that it does not seem to be possible to solve the race. 

When it comes to the impact, CA certificate change on disk while processing TLS handshake could cause connection failure, maybe even crash? `tls.Config.RootCAs` gets used at every reconnect, during processing of TLS handshake messages ([here](https://github.com/golang/go/blob/20e9b7f1b53d49fd66e0344b1d0d42d3cf5e47b6/src/crypto/tls/handshake_client.go#L862)). Similarly for servers, `tls.Config.ClientCAs` gets used when processing the TLS handshake ([here](https://github.com/golang/go/blob/20e9b7f1b53d49fd66e0344b1d0d42d3cf5e47b6/src/crypto/tls/handshake_server.go#L812)). The pointer to `CertPool` gets used when verifying the peer certificate during call `certs[0].Verify(opts)` (defined [here](https://github.com/golang/go/blob/20e9b7f1b53d49fd66e0344b1d0d42d3cf5e47b6/src/crypto/x509/verify.go#L751)).

Another observation: `certPool.AppendCertsFromPEM()` only appends new CA certificates. The old ones are never removed even if they would not be present on disk anymore. There isn't support in Go to remove certificates from an existing `CertPool`. 

In practise, it is quite common that only client and server credentials are rotated at runtime but modifying trust anchors require restart. It seems to me that the design of go crypto is (mostly) aligned with this.

BTW for TLS servers, there is callback `tls.Config.GetConfigForClient` that in theory could be used to return `tls.Config` with a different (atomic) copy of `tls.Config.ClientCAs`. There is no similar function for TLS clients.

Reply:
I booked #4316